src/pip/_internal/commands/install.py

```diff
@@ -546,63 +536,83 @@ def site_packages_writable(**kwargs):
+        if os.path.exists(target_dir) and not os.path.isdir(target_dir):
+            raise InstallationError("Target path exists but is not "
+                                    "a directory: {}".format(target_dir))
+        if not test_writable_dir(target_dir):
             raise InstallationError(
-                "Can not perform a '--user' install. User site-packages "
-                "are not visible in this virtualenv."
-            )
-        logger.debug("User install by explicit request")
-        return True
-
-    # If we are here, user installs have not been explicitly requested/avoided
-    assert use_user_site is None
-
-    # user install incompatible with --prefix/--target
-    if prefix_path or target_dir:
-        logger.debug("Non-user install due to --prefix or --target option")
+                "Cannot write to target directory: {}".format(target_dir))
+        logger.debug("Non-user install due to specified target directory")
         return False
-
-    # If user installs are not enabled, choose a non-user install
-    if not site.ENABLE_USER_SITE:
-        logger.debug("Non-user install because user site-packages disabled")
+    if prefix_path:
+        if not test_writable_dir(prefix_path):
+            raise InstallationError(
+                "Cannot write to prefix path: {}".format(prefix_path))
+        logger.debug("Non-user install due to specified prefix path")
         return False
 
-    # If we have permission for a non-user install, do that,
-    # otherwise do a user install.
-    if site_packages_writable(root=root_path, isolated=isolated_mode):
+    if use_user_site is not None:
+        use_user_site = bool(use_user_site)
+        logger.debug("{} install by explicit request".format(
+            "User" if use_user_site else "Non-user"))
+        if use_user_site is True and site.ENABLE_USER_SITE is None:
+            raise InstallationError("User site-packages are disabled for "
+                                    "security reasons or by an administrator.")
+    elif site.ENABLE_USER_SITE is None:
+        logger.debug("Non-user install since user site-packages are disabled "
+                     "for security reasons or by an administrator.")
+        use_user_site = False
+    elif site.ENABLE_USER_SITE is False:
+        logger.debug("Non-user install since user site-packages are disabled "
+                     "by user request (with 'python -s' or PYTHONNOUSERSITE)")
+        use_user_site = False
+    elif root_path:
+        logger.debug(
+            "Non-user install since alternate root directory is specified")
+        use_user_site = False
+    elif site_packages_writable(root=root_path, isolated=isolated_mode):
         logger.debug("Non-user install because site-packages writeable")
         return False
+    else:
+        logger.info("Defaulting to user installation because "
+                    "normal site-packages is not writeable")
+        use_user_site = True
 
```

Comment:
I suggest you just insert this assertion instead of `assert use_user_site is False` just in one branch.
```suggestion

    assert isinstance(use_user_site, bool)

```